### ISSUE TITLE
[SPMD] SPMD-DP with gradient accumulation does not reduce the communication across DP rank.

### ISSUE THREAD

**Issue by baoleai** [pytorch/xla#5760]
## ❓ Questions and Help
When we use SPMD for data parallelism and combine it with gradient accumulation(ga), we have noticed that the gradient all-reduce operation caused by DP appears in every step. However, the correct way is for the all-reduce to occur only when `step % ga == 0`. Do you have any suggestions for solving this issue? @JackCaoG @alanwaketan 

test script:

```
export XLA_USE_SPMD=1

WORLD_SIZE=1
RANK=0
MASTER_ADDR=127.0.0.1
MASTER_PORT=12456
torchrun --nproc_per_node=4 examples/test_spmd_dp.py\
        --nnodes $WORLD_SIZE\
        --node_rank $RANK \
        --master_addr $MASTER_ADDR \
        --master_port $MASTER_PORT \
        --log_interval 1 2>&1 | tee ./log/test.log
```

test_spmd_dp.py:
```
import numpy as np
import torch
import torch.utils.checkpoint
import torch_xla.core.xla_model as xm
import torch_xla.experimental.xla_sharding as xs
from torch import nn
from torch_xla.amp import syncfree
from torch_xla.experimental.xla_sharding import Mesh

num_devices = 4
devices_ids = np.arange(num_devices)
row_mesh = Mesh(devices_ids, (num_devices, 1), ('x', 'y'))


class SimpleLinear(nn.Module):

  def __init__(self):
    super(SimpleLinear, self).__init__()
    self.fc0 = nn.Linear(1024, 1024, bias=False)
    self.fc1 = nn.Linear(1024, 1024, bias=False)
    self.fc2 = nn.Linear(1024, 1024, bias=False)
    self.fc3 = nn.Linear(1024, 1024, bias=False)
    self.relu = nn.ReLU()


  def forward(self, x):
    h = self.relu(torch.einsum('ij,jk->ik', x, self.fc0.weight.T))
    y = self.relu(torch.einsum('ij,jk->ik', h, self.fc1.weight.T))
    z = torch.einsum('ij,jk->ik', y, self.fc3.weight.T)
    return z

ga = 4

def run():
    device = xm.xla_device()
    model = SimpleLinear().to(device)
    optimizer = syncfree.AdamW(model.parameters(), lr=0.0)

    with torch.profiler.profile(
            schedule=torch.profiler.schedule(
                wait=1, warmup=1, active=10),
            on_trace_ready=torch.profiler.tensorboard_trace_handler('test_prof')) as prof:
      for step in range(20):
          dummy_data = torch.zeros(64, 1024, device=device)
          xs.mark_sharding(dummy_data, row_mesh, (0, 1))
          x = dummy_data
          x = model(x)
          dummy_loss = x.sum() / ga
          dummy_loss.backward()
          if step % ga == 0: # grad all-reduce should only occur here.
              optimizer.step()
              optimizer.zero_grad()
          xm.mark_step()
          prof.step()

if __name__ == "__main__":
    run()
```

Accordding to profiling timeline, we find that the gradient all-reduce op caused by DP appears in every step rather than only on `step%ga == 0`.

**Comment by baoleai** [comment]
One possible fix is to shard the weight‘s gradient using sharding spec like {device_ids=[1,1,4], last_tile_dim_replicate=True}, so that the loss.backward() process does not insert an all-reduce. However, it seems that we cannot directly use the `mark_sharding` method to add such sharding spec.

**Comment by baoleai** [comment]
Maybe this issue is more related to openxla. https://github.com/openxla/xla/issues/6750

**Comment by alanwaketan** [comment]
Sorry for late reply. What's the behavior if you completely omit the whole optimizer step?

**Comment by baoleai** [comment]
In fact, we have not completely eliminated the optimize step. Instead, we perform the optimizer step after accumulating gradients for `GA_NUM`(ga=4 in above example) iterations, which is the correct logic for gradient accumulation. The current issue is that DP communication is also included in non-optimizer step iterations. However, these iterations (iteration_step % `GA_NUM` != 0) only require local gradient accumulation. DP communication and parameter updates are performed only in the iterations that require the optimizer step (iteration_step%`GA_NUM`==0).

**Comment by alanwaketan** [comment]
@baoleai That part I understand. I'm just curious if you see the gradient syncing even without the optimizer step.

**Comment by baoleai** [comment]
Yes, in the current version of XLA, I did observe gradient synchronization (all-reduce) in places without optimizer steps, as seen in the example above. This is because during SPMD pass, all-reduce is inserted at the backward dot operation. I will submit a PR to fix this issue in OpenXLA.

**Comment by alanwaketan** [comment]
> Yes, in the current version of XLA, I did observe gradient synchronization (all-reduce) in places without optimizer steps, as seen in the example above. This is because during SPMD pass, all-reduce is inserted at the backward dot operation. I will submit a PR to fix this issue in OpenXLA.

That's fantastic! Good luck on that as it could break a tons of things...
